From 2f7a1f729bd8d763e3fb7192cf84e3d27c1b7a47 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Thu, 29 Apr 2021 21:49:12 +0300
Subject: [PATCH] [SW-25959] greco: add firmware hard reset support

Add FW hard reset sequence flow to fix the rare bug when a transaction
is stuck in the PCIE_IF and prevents the chip from normal execution
after the reset.

As a part of the new reset flow the PCIe is closed for accesses in both,
the memory space access and the ELBI access. Therefore the new reset
sequence depends on the fact that after the reset Ppboot knows to open
the PCIe for requests.

Note that hard reset by firmware is not enabled, it will be enabled
once LKD adds hard reset by firmware support for greco.

Change-Id: I61388e3e011cd25c98f4e8affacc28a8627b0723
---

diff --git a/greco/Makefile b/greco/Makefile
index 4bfd3f5..73716ed 100644
--- a/greco/Makefile
+++ b/greco/Makefile
@@ -2,6 +2,7 @@
 PPBOOT_MAX_SIZE := 18432
 PREBOOT_SRCS_C += $(shell ls preboot/thermal_v2.c)
 PREBOOT_SRCS_C += $(shell ls $(HL_DEV)/security.c)
+PREBOOT_SRCS_C += $(shell ls $(HL_DEV)/reset.c)
 
 CFLAGS += -mstrict-align
 
diff --git a/greco/asic.c b/greco/asic.c
index 3bc0151..a94a3f1 100644
--- a/greco/asic.c
+++ b/greco/asic.c
@@ -84,7 +84,6 @@
 	struct hl_pci_inbound_region inbound_region;
 	struct hl_pci_outbound_region outbound_region;
 
-	/* TODO: implement hl_pcie_disable_access() for greco */
 	hl_pcie_disable_access();
 
 	/* Inbound Region 0 - Bar - - Point to SRAM and CFG */
@@ -104,7 +103,6 @@
 	outbound_region.size = HOST_PHYS_SIZE;
 	hl_pci_set_outbound_region(&outbound_region);
 
-	/* TODO: implement hl_pcie_enable_access() for greco */
 	hl_pcie_enable_access();
 }
 
diff --git a/greco/common.c b/greco/common.c
index 604572a..9c514a7 100644
--- a/greco/common.c
+++ b/greco/common.c
@@ -25,6 +25,7 @@
 #include <psoc_arc0_cfg_masks.h>
 #include <psoc_arc1_cfg_regs.h>
 #include <psoc_arc1_cfg_masks.h>
+#include <pcie_dbi_masks.h>
 
 #include <hl_boot_if.h>			 /* requires stdint.h */
 #include <greco/greco_reg_map.h>
@@ -694,3 +695,35 @@
 {
 	return (u32)mmPCIE_DBI_BASE & 0xFFFFFF;
 }
+
+/*
+ * This function makes sure no access is done from host to the SoC's PCIe IF.
+ * The inbound memory requests are disabled. The ELBI access gets "retry"
+ * response. All channels provide responses to the host and PCIe
+ * link will not break.
+ */
+void hl_pcie_disable_access(void)
+{
+	u32 reg;
+
+	/* Disable the ELBI access */
+	regmap_writel(1, mmPCIE_CORE_REQ_RETRY_EN);
+
+	/* Disable the PCIe inbound memory requests */
+	reg = regmap_readl(mmPCIE_DBI_STATUS_COMMAND_REG);
+	reg &= ~BIT(PCIE_DBI_STATUS_COMMAND_REG_PCI_TYPE0_MEM_SPACE_EN_SHIFT);
+	regmap_writel(reg, mmPCIE_DBI_STATUS_COMMAND_REG);
+}
+
+void hl_pcie_enable_access(void)
+{
+	u32 reg;
+
+	/* Enable the ELBI access */
+	regmap_writel(0, mmPCIE_CORE_REQ_RETRY_EN);
+
+	/* Enable the PCIe inbound memory requests */
+	reg = regmap_readl(mmPCIE_DBI_STATUS_COMMAND_REG);
+	reg |= BIT(PCIE_DBI_STATUS_COMMAND_REG_PCI_TYPE0_MEM_SPACE_EN_SHIFT);
+	regmap_writel(reg, mmPCIE_DBI_STATUS_COMMAND_REG);
+}
diff --git a/greco/reset.c b/greco/reset.c
new file mode 100644
index 0000000..cd9e761
--- /dev/null
+++ b/greco/reset.c
@@ -0,0 +1,41 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright 2021 HabanaLabs, Ltd.
+ * All Rights Reserved.
+ */
+
+#include <psoc_global_conf_regs.h>
+#include <psoc_global_conf_masks.h>
+
+#include <bitops.h>
+#include <preboot.h>
+#include <pll.h>
+#include <uart.h>
+
+#include "common.h"
+
+#define SW_ALL_RST_L_MSK		0xC00FF8F2
+#define SW_ALL_RST_M_MSK		0xFFFFFFFF
+#define SW_ALL_RST_H_MSK		0x7FFFFF
+
+static void greco_hw_fini(void)
+{
+	/* Restart BTL/BLR upon hard-reset */
+	regmap_writel(0x770, mmPSOC_GLOBAL_CONF_BOOT_SEQ_RE_START);
+
+	udelay(1000);
+
+	regmap_writel(SW_ALL_RST_L_MSK, mmPSOC_GLOBAL_CONF_SW_ALL_RST_CFG_L);
+	regmap_writel(SW_ALL_RST_M_MSK, mmPSOC_GLOBAL_CONF_SW_ALL_RST_CFG_M);
+	regmap_writel(SW_ALL_RST_H_MSK, mmPSOC_GLOBAL_CONF_SW_ALL_RST_CFG_H);
+
+	regmap_writel(1, mmPSOC_GLOBAL_CONF_SW_ALL_RST);
+}
+
+void hl_dev_reset(void)
+{
+	hl_pcie_disable_access();
+
+	greco_hw_fini();
+	wfe();
+}
